File: frontend/src/lib/scenario.ts
/**
 * Pure helpers for the user-weight scenario lab (가중치 실험실), extracted so the
 * weight math, normalization, rank-movement text, and session persistence are
 * unit-testable independently of the React tree (which Playwright exercises).
 *
 * The editor works in an integer 0–100 percent scale; the API receives canonical
 * 8-dp decimal strings. Integer percents that total exactly 100 map to canonical
 * weights that sum to exactly 1.00000000, so a valid editor total always yields a
 * valid canonical vector. Nothing here silently repairs an invalid total — only
 * the explicit "100%로 비율 정규화" action changes values, and it reports that it did.
 */

import type { SuitabilityProfile, SuitabilityRun, UserScenarioWeights } from "./api";

export const SCENARIO_COMPONENTS = ["zoning", "road", "equity", "demand"] as const;
export type ScenarioComponent = (typeof SCENARIO_COMPONENTS)[number];
export type ScenarioPercents = Record<ScenarioComponent, number>;

/** Citizen-facing labels + component code for each weight control. */
export const SCENARIO_COMPONENT_META: Record<
  ScenarioComponent,
  { label: string; code: "Z" | "R" | "E" | "D"; explanation: string }
> = {
  zoning: { label: "토지이용", code: "Z", explanation: "용도지역 적합도 (land-use context)" },
  road: { label: "도로 접근성", code: "R", explanation: "도로 근접 대리지표 (access proxy)" },
  equity: { label: "형평성", code: "E", explanation: "기존 시설 부담 회피 (burden avoidance)" },
  demand: { label: "처리 수요", code: "D", explanation: "1인당 처리 수요 (per-capita demand)" },
};

/** Preset weight sources available in a run, with citizen-facing labels. */
export const SCENARIO_PRESET_LABELS: Record<string, string> = {
  baseline: "기본 가정",
  equal: "균등",
  equity_focused: "형평성 중심",
  access_focused: "접근성 중심",
  critic: "CRITIC 데이터 기반",
};

export const SCENARIO_STORAGE_KEY = "waste-equity:suitability-scenario:v1";
export const SCENARIO_SESSION_SCHEMA = 1;

/** Sum of the four percents (may be ≠ 100 while the user is editing). */
export function draftTotal(percents: ScenarioPercents): number {
  return SCENARIO_COMPONENTS.reduce((acc, c) => acc + percents[c], 0);
}

/** A draft is applicable only when every value is in [0,100] and the total is exactly 100. */
export function isDraftValid(percents: ScenarioPercents): boolean {
  return (
    SCENARIO_COMPONENTS.every(
      (c) => Number.isInteger(percents[c]) && percents[c] >= 0 && percents[c] <= 100,
    ) && draftTotal(percents) === 100
  );
}

/** Difference from 100 (positive = over, negative = under). */
export function totalDifference(percents: ScenarioPercents): number {
  return draftTotal(percents) - 100;
}

/**
 * Canonical 8-dp weight strings from integer percents. Requires a valid total of
 * exactly 100 (each p/100 is exact to 2 dp, so the sum is exactly 1.00000000).
 * Throws on an invalid total rather than silently normalizing.
 */
export function percentsToCanonical(percents: ScenarioPercents): UserScenarioWeights {
  if (!isDraftValid(percents)) {
    throw new Error("percentsToCanonical requires integer percents summing to exactly 100");
  }
  const toStr = (p: number): string => (p / 100).toFixed(8);
  return {
    zoning: toStr(percents.zoning),
    road: toStr(percents.road),
    equity: toStr(percents.equity),
    demand: toStr(percents.demand),
  };
}

/**
 * Deterministic largest-remainder allocation of `total` across the four components
 * proportional to `raw`, producing integers that sum exactly to `total`. Ties in
 * the fractional part break by fixed Z/R/E/D order. Returns null when every raw
 * value is zero (nothing to allocate).
 */
function largestRemainder(raw: number[], total: number): number[] | null {
  const sum = raw.reduce((a, b) => a + b, 0);
  if (sum <= 0) return null;
  const scaled = raw.map((v) => (v / sum) * total);
  const floored = scaled.map(Math.floor);
  let remainder = total - floored.reduce((a, b) => a + b, 0);
  const order = scaled
    .map((v, i) => ({ i, frac: v - Math.floor(v) }))
    .sort((a, b) => b.frac - a.frac || a.i - b.i);
  const out = [...floored];
  for (let k = 0; k < remainder && k < order.length; k++) out[order[k].i] += 1;
  // Any residual (only possible from float underflow) goes to the first slot.
  remainder = total - out.reduce((a, b) => a + b, 0);
  if (remainder !== 0) out[0] += remainder;
  return out;
}

/**
 * Explicit "100%로 비율 정규화": rescale the current draft to sum exactly 100.
 * Returns the new percents and whether any value changed, or null when all values
 * are zero (normalization is undefined — the caller must inform the user).
 */
export function normalizePercents(
  percents: ScenarioPercents,
): { percents: ScenarioPercents; changed: boolean } | null {
  const raw = SCENARIO_COMPONENTS.map((c) => percents[c]);
  const allocated = largestRemainder(raw, 100);
  if (allocated === null) return null;
  const next: ScenarioPercents = {
    zoning: allocated[0],
    road: allocated[1],
    equity: allocated[2],
    demand: allocated[3],
  };
  const changed = SCENARIO_COMPONENTS.some((c) => next[c] !== percents[c]);
  return { percents: next, changed };
}

/**
 * Integer percents from a stored decimal weight profile (e.g. {"zoning":"0.35"}),
 * normalized to sum exactly 100 by largest remainder so a preset is always a valid
 * loadable draft. Used only for preset buttons — never fabricates a missing profile.
 */
export function decimalWeightsToPercents(profile: Record<string, string>): ScenarioPercents {
  const raw = SCENARIO_COMPONENTS.map((c) => Number(profile[c] ?? "0") * 100);
  const allocated = largestRemainder(raw, 100) ?? [25, 25, 25, 25];
  return {
    zoning: allocated[0],
    road: allocated[1],
    equity: allocated[2],
    demand: allocated[3],
  };
}

/**
 * Preset weight sources actually available for a run, in a stable order, as
 * loadable integer percents. CRITIC appears only when the run computed it (present
 * in weight_profiles) — never fabricated for old runs.
 */
export function scenarioPresets(
  run: SuitabilityRun | null | undefined,
): { key: string; label: string; percents: ScenarioPercents }[] {
  const profiles = run?.weight_profiles ?? {};
  const order = ["baseline", "equal", "equity_focused", "access_focused", "critic"];
  const out: { key: string; label: string; percents: ScenarioPercents }[] = [];
  for (const key of order) {
    const profile = profiles[key];
    if (!profile) continue;
    out.push({
      key,
      label: SCENARIO_PRESET_LABELS[key] ?? key,
      percents: decimalWeightsToPercents(profile),
    });
  }
  return out;
}

/**
 * Text-first rank movement (never color-only). Examples:
 *   42위 → 18위, 24계단 상승 / 18위 → 42위, 24계단 하락 / 순위 변화 없음.
 * Returns a neutral string when a comparison rank is unavailable.
 */
export function rankMovementText(
  comparisonRank: number | null | undefined,
  customRank: number | null | undefined,
): string {
  if (customRank == null) return "순위 없음";
  if (comparisonRank == null) return `${customRank}위 (비교 순위 없음)`;
  const delta = comparisonRank - customRank;
  if (delta === 0) return "순위 변화 없음";
  const dir = delta > 0 ? "상승" : "하락";
  return `${comparisonRank}위 → ${customRank}위, ${Math.abs(delta)}계단 ${dir}`;
}

// --------------------------------------------------------------------------- //
// Session-only persistence (sessionStorage — never localStorage/cookies/URL).
// --------------------------------------------------------------------------- //

export interface ScenarioSessionState {
  schemaVersion: number;
  runId: number;
  draftPercents: ScenarioPercents;
  appliedPercents: ScenarioPercents | null;
  compareProfile: SuitabilityProfile;
  scenarioHash: string | null;
  selectedCandidateId: number | null;
}

function isPercents(value: unknown): value is ScenarioPercents {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return SCENARIO_COMPONENTS.every(
    (c) => typeof v[c] === "number" && Number.isFinite(v[c] as number),
  );
}

/** Persist scenario UI state for this browser tab only. No-op without sessionStorage. */
export function saveScenarioSession(state: ScenarioSessionState): void {
  if (typeof sessionStorage === "undefined") return;
  try {
    sessionStorage.setItem(SCENARIO_STORAGE_KEY, JSON.stringify(state));
  } catch {
    /* storage full / disabled — session restore is best-effort only */
  }
}

export function clearScenarioSession(): void {
  if (typeof sessionStorage === "undefined") return;
  try {
    sessionStorage.removeItem(SCENARIO_STORAGE_KEY);
  } catch {
    /* ignore */
  }
}

/**
 * Load + validate stored scenario state for the CURRENT run. Returns null (and the
 * caller ignores/clears it) when the schema is wrong, weights are invalid, the run
 * id changed, or the comparison profile is no longer available — the restored draft
 * is only ever shown after a fresh preview request re-verifies it.
 */
export function loadScenarioSession(
  runId: number,
  availableProfiles: readonly SuitabilityProfile[],
): ScenarioSessionState | null {
  if (typeof sessionStorage === "undefined") return null;
  let raw: string | null = null;
  try {
    raw = sessionStorage.getItem(SCENARIO_STORAGE_KEY);
  } catch {
    return null;
  }
  if (!raw) return null;
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    return null;
  }
  if (typeof parsed !== "object" || parsed === null) return null;
  const s = parsed as Record<string, unknown>;
  if (s.schemaVersion !== SCENARIO_SESSION_SCHEMA) return null;
  if (s.runId !== runId) return null; // active run changed → discard
  if (!isPercents(s.draftPercents)) return null;
  if (s.appliedPercents !== null && !isPercents(s.appliedPercents)) return null;
  if (
    typeof s.compareProfile !== "string" ||
    !availableProfiles.includes(s.compareProfile as SuitabilityProfile)
  ) {
    return null;
  }
  return {
    schemaVersion: SCENARIO_SESSION_SCHEMA,
    runId,
    draftPercents: s.draftPercents as ScenarioPercents,
    appliedPercents: (s.appliedPercents as ScenarioPercents | null) ?? null,
    compareProfile: s.compareProfile as SuitabilityProfile,
    scenarioHash: typeof s.scenarioHash === "string" ? s.scenarioHash : null,
    selectedCandidateId:
      typeof s.selectedCandidateId === "number" ? s.selectedCandidateId : null,
  };
}

File: backend/src/waste_equity_backend/analysis/suitability/scenario.py
"""User-defined weight *scenario* domain logic (Phase 6).

A user-weight scenario is a **temporary decision-support experiment**: it takes
the four frozen component scores (zoning / road / equity / demand) of one fixed,
already-succeeded :class:`SuitabilityAnalysisRun` and *recombines* them under a
user-supplied weight vector, on read, without ever writing to the database.

It is emphatically **not**:

* an official suitability profile (``baseline``/``equal``/``equity_focused``/
  ``access_focused``/``critic`` are the only stored profiles),
* an analytical run (nothing is persisted to ``suitability_analysis_runs`` or
  ``suitability_candidates``),
* part of CRITIC derivation or stored stability classification,
* a legal, engineering, environmental-review, permitting, or final-siting result.

Because it only *reweights frozen stored scores*, it introduces **no new stored
derivation** and therefore its own method version — :data:`USER_WEIGHT_SCENARIO_METHOD_VERSION`
— is independent of and does **not** bump ``suitability-policy-v2``,
``suitability-screening-v3``, ``critic-weights-v1``, ``suitability-stability-v1``,
or ``capital-grid-500m-v1``.

Everything here is pure and independently testable: parsing, validation,
quantization, canonical serialization, the exact score/provisional formulas, the
deterministic scenario hash, and the rank-delta convention. The scoring math
reuses :mod:`policy` (``composite`` / ``provisional_composite`` /
``quantize_score``) so the Python helper, the preview SQL, the candidate-detail
SQL, and the MVT SQL all agree byte-for-byte on the same 0–100, 4-dp, ROUND_HALF_EVEN
scale. See ``docs/SUITABILITY_USER_WEIGHT_SCENARIOS.md``.
"""

from __future__ import annotations

import hashlib
import json
from collections.abc import Mapping, Sequence
from decimal import ROUND_HALF_EVEN, Decimal, InvalidOperation
from typing import Any

from . import component_model, policy

# --------------------------------------------------------------------------- #
# Method version (independent of stored-run derivation versions)
# --------------------------------------------------------------------------- #

# A user scenario reweights *frozen stored scores* on read. It changes no stored
# run derivation or candidate classification, so it carries its own, separate
# method version and never increments the run/policy/critic/stability/grid
# versions. Bumping this only signals a change to the *scenario recombination
# contract itself* (weight model, hashing payload, or scoring/quantization).
#
# Deliberately NOT bumped by making weight validation run-model-relative. For every
# run that can exist today the rule resolves to exactly the historical four keys, so
# no producible scenario's weight model, hash payload, or quantization has changed,
# and every stored scenario hash stays byte-identical. The bump to
# ``user-weight-scenario-v2`` — together with adding the component model to
# :func:`canonical_hash_payload` so a successor scenario hash can never collide with
# a historical one — is required at the moment successor scenarios become
# producible, i.e. when the successor run write path and an approved successor
# weight vector land. Until then a bump would change every existing hash for a
# contract change that has not actually happened.
USER_WEIGHT_SCENARIO_METHOD_VERSION = "user-weight-scenario-v1"

# Fixed criterion order for the *historical* scenario weight vector, hash payload,
# and serialization (identical to policy.COMPONENTS / critic.CRITERION_ORDER). It is
# the default for every function below; a run of another component model supplies
# its own ``component_order``, which is why none of them read this constant when an
# order is passed in.
COMPONENT_ORDER: tuple[str, ...] = policy.COMPONENTS

# Canonical scenario-weight precision: 8 decimal places (matches the CRITIC
# vector precision so a CRITIC preset round-trips exactly).
WEIGHT_QUANT = Decimal("0.00000001")  # 8 dp
_ZERO = Decimal("0")
_ONE = Decimal("1")
_CANONICAL_ONE = Decimal("1.00000000")

# Structured error code surfaced as a 422 by the API layer.
INVALID_SCENARIO_WEIGHTS = "INVALID_SCENARIO_WEIGHTS"

# A weight vector that is well-formed for a *different* component model gets its own
# code, deliberately distinct from INVALID_SCENARIO_WEIGHTS, because the remedy is
# completely different: malformed weights are a correctable input error, whereas a
# model mismatch means this scenario cannot be applied to this run at all. Collapsing
# the two would push a client toward "fix your weights", which is the wrong
# instruction and nudges precisely toward remapping one model's weights onto
# another's components.
COMPONENT_MODEL_MISMATCH = component_model.COMPONENT_MODEL_MISMATCH

# Citizen-facing scenario label + disclaimer (kept here so backend responses and
# the docs share one source of truth; the frontend mirrors these strings).
SCENARIO_LABEL_KO = "사용자 가정 기반 시나리오"
SCENARIO_DISCLAIMER_KO = (
    "사용자가 입력한 가중치로 기존 분석 실행의 Z/R/E/D 구성점수를 재결합한 임시 비교 "
    "결과입니다. 공식 분석 실행, 전문가 판단, 법적 적격성, 인허가 가능성 또는 최종 입지 "
    "결정을 의미하지 않습니다."
)


class ScenarioWeightError(ValueError):
    """Invalid user scenario weights.

    ``error`` is a stable machine code (``INVALID_SCENARIO_WEIGHTS``); ``detail``
    is a human-readable message; ``fields`` carries structured context (e.g. the
    offending canonical ``sum``). The API layer maps this to a 422 body::

        {"error": ..., "detail": ..., "fields": {...}}

    Invalid weights are **never** silently normalized, replaced with equal
    weights, or have a remainder redistributed — the caller is always informed.
    """

    error = INVALID_SCENARIO_WEIGHTS

    def __init__(self, detail: str, fields: dict[str, Any] | None = None) -> None:
        super().__init__(detail)
        self.detail = detail
        self.fields = fields or {}

    def as_envelope(self) -> dict[str, Any]:
        return {"error": self.error, "detail": self.detail, "fields": self.fields}


class ScenarioComponentModelMismatchError(ScenarioWeightError):
    """Well-formed weights for one component model, submitted against another's run.

    A subclass of :class:`ScenarioWeightError` so every existing ``except`` clause
    still catches it and still produces a 422, but with its own stable ``error``
    code so a client can present "this scenario belongs to a different analytical
    model" instead of "your weights are wrong". The scenario is not broken — it is
    valid for a different model, and remains fully usable against a run of that
    model.
    """

    error = COMPONENT_MODEL_MISMATCH


def _parse_one(component: str, raw: Any) -> Decimal:
    """Parse a single weight to an exact finite Decimal (never binary float).

    Strings are preferred and parsed with :class:`Decimal` directly. A ``float``
    is rejected: JSON binary floating-point must not silently enter the canonical
    weight math. NaN / Infinity / malformed input raise
    :class:`ScenarioWeightError`.
    """

    if isinstance(raw, bool):  # bool is an int subclass; never a weight
        raise ScenarioWeightError(
            f"Weight '{component}' must be a decimal string, not a boolean.",
            {"component": component},
        )
    if isinstance(raw, float):
        raise ScenarioWeightError(
            f"Weight '{component}' must be a decimal string, not a binary float "
            "(floating-point values are not accepted for canonical weights).",
            {"component": component},
        )
    if isinstance(raw, Decimal):
        value = raw
    elif isinstance(raw, int):
        value = Decimal(raw)
    elif isinstance(raw, str):
        text = raw.strip()
        if not text:
            raise ScenarioWeightError(f"Weight '{component}' is empty.", {"component": component})
        try:
            value = Decimal(text)
        except InvalidOperation as exc:
            raise ScenarioWeightError(
                f"Weight '{component}' is not a valid decimal: {raw!r}.",
                {"component": component},
            ) from exc
    else:
        raise ScenarioWeightError(
            f"Weight '{component}' must be a decimal string.", {"component": component}
        )
    if value.is_nan() or value.is_infinite():
        raise ScenarioWeightError(
            f"Weight '{component}' must be finite (got {raw!r}).", {"component": component}
        )
    return value


def parse_and_validate_weights(
    raw: Mapping[str, Any], components: Sequence[str] | None = None
) -> dict[str, Decimal]:
    """Validate a raw weight mapping and return canonical 8-dp Decimal weights.

    ``components`` is the component order the weights must be defined over — pass
    the *resolved run's* ``component_order`` so validation is run-model-relative
    rather than relative to a module constant. It defaults to the historical
    :data:`COMPONENT_ORDER`, which is what every existing caller already meant.

    Enforces, with no silent repair:

    * exactly the model's component keys, no missing and no unknown keys;
    * each value a finite decimal in ``[0, 1]`` inclusive (zero allowed);
    * not all zero;
    * the canonical (8-dp-quantized) sum equals exactly ``Decimal("1.00000000")``.

    Strictness here is the load-bearing safety property of the whole scenario path:
    it is the only thing standing between a saved weight vector and a silent
    cross-model recombination. It must never be loosened to "accept any subset and
    renormalize".

    A key set that is *exactly another known component model's* components raises
    :class:`ScenarioComponentModelMismatchError` (a ``ScenarioWeightError`` subclass
    carrying the distinct ``COMPONENT_MODEL_MISMATCH`` code), so a client can tell
    "this scenario belongs to a different analytical model" apart from "your weights
    are malformed". Anything else raises :class:`ScenarioWeightError` (→ 422) as
    before. Prefer passing decimal strings; floats are rejected upstream in
    :func:`_parse_one`.
    """

    order = tuple(COMPONENT_ORDER if components is None else components)
    if not isinstance(raw, Mapping):
        raise ScenarioWeightError(
            f"Scenario weights must be an object of {len(order)} components."
        )
    keys = set(raw)
    expected = set(order)
    missing = expected - keys
    unknown = keys - expected
    if missing or unknown:
        # Well-formed weights for a different component model are refused with a
        # distinct code and are NEVER translated: the components are different
        # measured quantities with different derivations, resolutions, and
        # directions, and a positional or name-similarity mapping would carry a
        # justification written about one quantity onto another.
        submitted_model = component_model.classify_weight_components(raw)
        if submitted_model is not None:
            raise ScenarioComponentModelMismatchError(
                "These scenario weights are defined over the components of "
                f"{submitted_model!r} ({sorted(keys)}) and cannot be applied to a run "
                f"whose components are {list(order)}. A scenario is only valid against "
                "a run of its own component model; weights are never remapped between "
                "models.",
                {
                    "submitted_component_model": submitted_model,
                    "submitted_components": sorted(keys),
                    "run_component_order": list(order),
                },
            )
    if missing:
        raise ScenarioWeightError(
            "Scenario weights must include exactly " + ", ".join(order) + ".",
            {"missing": sorted(missing), "expected": list(order)},
        )
    if unknown:
        raise ScenarioWeightError(
            "Scenario weights contain unknown components.",
            {"unknown": sorted(unknown), "expected": list(order)},
        )

    parsed = {c: _parse_one(c, raw[c]) for c in order}
    for c, value in parsed.items():
        if value < _ZERO or value > _ONE:
            raise ScenarioWeightError(
                f"Weight '{c}' must be between 0 and 1 inclusive (got {value}).",
                {"component": c, "value": format(value, "f")},
            )

    canonical = {c: parsed[c].quantize(WEIGHT_QUANT, rounding=ROUND_HALF_EVEN) for c in order}

    if all(v == _ZERO for v in canonical.values()):
        raise ScenarioWeightError(
            "Scenario weights cannot all be zero.",
            {"sum": format(_ZERO.quantize(WEIGHT_QUANT), "f")},
        )

    total = sum(canonical.values(), start=_ZERO)
    if total != _CANONICAL_ONE:
        raise ScenarioWeightError(
            "Scenario weights must sum exactly to 1.00000000.",
            {"sum": format(total, "f")},
        )
    return canonical


def canonical_weight_strings(
    weights: Mapping[str, Decimal], components: Sequence[str] | None = None
) -> dict[str, str]:
    """Fixed-point 8-dp strings in fixed criterion order (never exponent form).

    ``components`` defaults to the historical :data:`COMPONENT_ORDER`; pass the
    run's ``component_order`` for any other model. Order is explicit rather than
    taken from the mapping's key order, because key order is not a stable property
    of anything that round-trips through JSON.
    """

    order = COMPONENT_ORDER if components is None else components
    return {
        c: format(weights[c].quantize(WEIGHT_QUANT, rounding=ROUND_HALF_EVEN), "f") for c in order
    }


def scenario_score(
    component_scores: Mapping[str, Decimal], weights: Mapping[str, Decimal]
) -> Decimal:
    """Exact custom composite for an ELIGIBLE candidate (all four components present).

    ``custom_score = Σ component_score · weight``, on the 0–100 scale, quantized to
    4 dp with ROUND_HALF_EVEN — identical to :func:`policy.composite`, so it matches
    every stored-composite score and the SQL scoring paths exactly.
    """

    return policy.composite(dict(component_scores), dict(weights))


def scenario_provisional_score(
    component_scores: Mapping[str, Decimal], weights: Mapping[str, Decimal]
) -> Decimal | None:
    """Provisional custom composite for a REVIEW_REQUIRED candidate.

    Renormalizes over the components actually present (missing components are
    never zero-filled). Returns ``None`` when no component is present or the total
    weight of the present components is zero. Identical semantics to
    :func:`policy.provisional_composite`.
    """

    return policy.provisional_composite(dict(component_scores), dict(weights))


# --------------------------------------------------------------------------- #
# Deterministic scenario identity
# --------------------------------------------------------------------------- #


def canonical_hash_payload(
    run_id: int, weights: Mapping[str, Decimal], components: Sequence[str] | None = None
) -> str:
    """The exact UTF-8 string that is SHA-256'd for the scenario hash.

    Fixed key order (``method_version``, ``run_id``, ``weights`` in criterion
    order), compact separators, no whitespace. Excludes selected candidate,
    top_n, viewport, comparison profile, timestamps, and any frontend label — so
    only ``(method_version, run_id, canonical weights)`` determine the identity.

    ``components`` defaults to the historical :data:`COMPONENT_ORDER`, exactly as
    :func:`canonical_weight_strings` does, so a historical scenario's payload — and
    therefore its hash — is byte-identical to what it has always been. Pass the
    run's ``component_order`` for any other model: without it a successor scenario's
    weights would be looked up under historical component names it does not carry.
    """

    payload = {
        "method_version": USER_WEIGHT_SCENARIO_METHOD_VERSION,
        "run_id": int(run_id),
        "weights": canonical_weight_strings(weights, components),
    }
    return json.dumps(payload, separators=(",", ":"), ensure_ascii=False)


def scenario_hash(
    run_id: int, weights: Mapping[str, Decimal], components: Sequence[str] | None = None
) -> str:
    """Deterministic SHA-256 (full 64-hex) identity of a scenario.

    Same run + same canonical weights → same hash; a different run id or any
    different weight → a different hash. The comparison profile never affects it.
    This is a *temporary analytical identity*, not a database id.

    ``components`` carries the run's component order for a non-historical model
    (see :func:`canonical_hash_payload`); omitted, the historical order is used and
    a historical scenario keeps the hash it has always had.
    """

    return hashlib.sha256(
        canonical_hash_payload(run_id, weights, components).encode("utf-8")
    ).hexdigest()


def short_scenario_hash(full_hash: str, length: int = 12) -> str:
    """A documented collision-resistant display prefix of the full hash.

    12 hex chars = 48 bits; used only for compact display. The full hash remains
    available and is what the MVT endpoint validates against.
    """

    return full_hash[:length]


# --------------------------------------------------------------------------- #
# Rank-delta convention
# --------------------------------------------------------------------------- #

RANK_UP = "up"
RANK_DOWN = "down"
RANK_SAME = "same"


def rank_delta(comparison_rank: int | None, custom_rank: int | None) -> int | None:
    """``rank_delta = comparison_profile_rank − custom_rank``.

    Positive → the candidate moved *up* under the custom scenario (better/lower
    rank number); zero → unchanged; negative → moved *down*. ``None`` when either
    rank is unavailable (e.g. a REVIEW_REQUIRED/EXCLUDED candidate has no rank).
    """

    if comparison_rank is None or custom_rank is None:
        return None
    return comparison_rank - custom_rank


def rank_change_direction(delta: int | None) -> str | None:
    """Text direction for a rank delta (never color-only): up / same / down."""

    if delta is None:
        return None
    if delta > 0:
        return RANK_UP
    if delta < 0:
        return RANK_DOWN
    return RANK_SAME
